_posts/2021/2021-03-31-hmat_memkind.md

```diff
@@ -0,0 +1,172 @@
+Instead of providing a single shared bus, the concept proposes a solution to divide the hardware resources into `NUMA Nodes` abstraction.
+Depending on the exact implementation on multi-socket platforms, NUMA Node could be exposed as CPU(s) and memory placed inside the same socket.
+Accessing the memory placed on the same socket as the cpu on which process runs (`local memory access`) is faster since it avoids
+additional overhead from interconnection bus, like the memory access between two sockets (`remote memory access`).
+
+<figure class="image">
+  <img src="/assets/numa_overview.png" alt="NUMA overview">
+  <figcaption>Figure 1. NUMA overview.</figcaption>
+</figure>
+
+The operating system made the memory allocation decision among other factors based on `NUMA distance.` The distance represents
+relative memory latency between NUMA nodes, and is based on hardware configuration - System Locality Information Table (SLIT) and
+System Resource Affinity Table (SRAT) presented in [ACPI][ACPI-doc].
+NUMA aware system Operating System can still consume all possible memory, but the OS will optimize it for local memory access.
+In practice, this means that memory allocation will come from the NUMA Node with the smallest distance from the CPU of the calling thread.
+
+NUMA memory access pattern could be decribed as **heterogenous memory access**.
+
+But today's NUMA Node concept is vague compared to the original one. Extension of different Processing Units
+provides **heterogenous computing**, but the memory doesn't stay behind in this race. The NUMA Node itself now
+doesn't have to be a couple between DRAM and CPU.
+For example, we could have a memory-only NUMA Node (hot-added memory device), the NUMA Node, which cannot
+generate memory requests and contains only memory.
+From a processing perspective, we could have a different type of memory initiators: today, the CPU can perform
+memory allocation request and other processing units as GPU or FPGA. The same applies to memory targets as this
+could be different memory physical mediums like DRAM, PMEM, and HBW.
+
+This post focuses on the last aspect of mentioned varieties - on **heterogenous memory**.
+
+<figure class="image">
+  <img src="/assets/heterogeneous_architecture.png" alt="PMEM KMEM-DAX">
+  <figcaption>Figure 2. Heterogeneous architecture overview.</figcaption>
+</figure>
+
+### Memory technology vs memory property
+
+Different physical medium offers other memory performance characteristics like latency, bandwidth or capacity depending on media type and memory bus.
+From the software layer, application requirements against tasks are often related not to the underlying hardware itself but to mitigate specific bottlenecks related to the memory property. To overcome bounds, we use memory-aware programming - this approach is already used in different [interfaces][openmp].
```

Comment:
this is hard to parse.

"related ... to mitigate" - this is not correct.

"overcome bounds" - odd phrasing

"which is an approached already in"